ext/opentelemetry-ext-asgi/src/opentelemetry/ext/asgi/__init__.py

```diff
@@ -0,0 +1,210 @@
+    method_or_path = scope.get("method") or scope.get("path")
+
+    return method_or_path
+
+
+class OpenTelemetryMiddleware:
+    """The ASGI application middleware.
+
+    This class is an ASGI middleware that starts and annotates spans for any
+    requests it is invoked with.
+
+    Args:
+        app: The ASGI application callable to forward requests to.
+        name_callback: Callback which calculates a generic span name for an
+                       incoming HTTP request based on the ASGI scope.
+                       Optional: Defaults to get_default_span_name.
+    """
+
+    def __init__(self, app, name_callback=None):
+        self.app = guarantee_single_callable(app)
+        self.tracer = trace.get_tracer(__name__, __version__)
+        self.name_callback = name_callback or get_default_span_name
+
+    async def __call__(self, scope, receive, send):
+        """The ASGI application
+
+        Args:
+            scope: A ASGI environment.
+            receive: An awaitable callable yielding dictionaries
+            send: An awaitable callable taking a single dictionary as argument.
+        """
+        if scope.get("type") not in ("http", "websocket"):
+            return await self.app(scope, receive, send)
+
+        token = context.attach(
+            propagators.extract(get_header_from_scope, scope)
+        )
+        span_name = self.name_callback(scope)
+
+        try:
+            with self.tracer.start_as_current_span(
+                span_name + " asgi",
+                kind=trace.SpanKind.SERVER,
+                attributes=collect_request_attributes(scope),
+            ):
+
+                @wraps(receive)
+                async def wrapped_receive():
+                    with self.tracer.start_as_current_span(
+                        span_name + " asgi." + scope["type"] + ".receive"
+                    ) as receive_span:
+                        payload = await receive()
+                        if payload["type"] == "websocket.receive":
+                            set_status_code(receive_span, 200)
+                        receive_span.set_attribute("type", payload["type"])
+                    return payload
+
+                @wraps(send)
+                async def wrapped_send(payload):
+                    with self.tracer.start_as_current_span(
+                        span_name + " asgi." + scope["type"] + ".send"
+                    ) as send_span:
+                        if payload["type"] == "http.response.start":
+                            status_code = payload["status"]
+                            set_status_code(send_span, status_code)
+                        elif payload["type"] == "websocket.send":
+                            set_status_code(send_span, 200)
+                        send_span.set_attribute("type", payload["type"])
+                        await send(payload)
+
+                await self.app(scope, wrapped_receive, wrapped_send)
```

Comment:
If OpenTelemetry supports some kind of "attach a traceback to a trace", then we could also `try/except` around the app call; otherwise, please ignore.

Reply:
The Python OpenTelemetry client does not store the traceback.

Reply:
I think this is still up for debate in the standard. But it's a good point, and something I know is highly valuable around existing DD integration. @majorgreys thoughts? should we raise this in the spec?